Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt

```diff
@@ -0,0 +1,327 @@
+// Speed of animation impacts the ranges we select. For example, we must have a larger range to
+// detect initial collision with floor as ball animates quickly through it, otherwise
+// it'd be possible for no recomposition to occur within range.
+private val DISTANCE_FROM_END_POSITION_FOR_COLLISION = 5.dp
+private val DISTANCE_FROM_START_POSITION_FOR_COLLISION = 1.dp
+
+private val FLOOR_SIZE = 80.dp
+
+private const val DROP_ANIMATION_TIME_MS = 3000
+private const val RESET_ANIMATION_TIME_MS = 1000
+
+@Composable
+fun BounceRoute(viewModel: BounceViewModel) {
+    BounceExampleScreen(messageToUser = viewModel.messageToUser)
+}
+
+@Composable
+private fun BounceExampleScreen(messageToUser: String) {
+    var ballState by remember { mutableStateOf(BallState.Start) }
+    var bounceCount by remember { mutableStateOf(0) }
+
+    val vibrator = LocalContext.current.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
+
+    var transitionData = updateTransitionData(
+        ballState = ballState
+    )
+    val collisionData = updateCollisionData(transitionData)
+
+    // Execute a click vibration once the ball has been reset.
+    var hasVibratedForReset by remember { mutableStateOf(false) }
+    if (collisionData.collisionWithReset) {
+        if (!hasVibratedForReset) {
+            clickVibration(vibrator)
+            hasVibratedForReset = true
+        }
+    }
+
+    // Ball is about to contact floor, only vibrating once per collision.
+    var hasVibratedForBallContact by remember { mutableStateOf(false) }
+    if (collisionData.collisionWithFloor) {
+        if (!hasVibratedForBallContact) {
+            thudVibration(vibrator, 0.7.pow(bounceCount++).toFloat())
+            hasVibratedForBallContact = true
+        }
+    } else {
+        // Reset for next contact with floor.
+        hasVibratedForBallContact = false
+    }
+
+    Screen(pageTitle = stringResource(R.string.bounce), messageToUser = messageToUser) {
+        Box(
+            Modifier
+                .fillMaxSize()
+                .noRippleClickable {
+                    if (transitionData.isAtStart) {
+                        ballState = BallState.End
+                    } else {
+                        // Reset the state, with a thud vibration to simulate bounce off the floor.
+                        ballState = BallState.Start
+                        bounceCount = 0
+                        hasVibratedForBallContact = false
+                        hasVibratedForReset = false
+                        thudVibration(vibrator)
+                    }
+                },
+        ) {
+
+            var instructionsText: String = "" // Don't display any instructions when bouncing.
+            if (transitionData.isAtStart) {
+                instructionsText = stringResource(R.string.bounce_tap_to_drop)
+            } else if (transitionData.isAtEnd || transitionData.isResetting) {
```

Comment:
Do you mean !transitionData.isResetting here? Or do you want to show only after the end of the bounce animation?